rclcpp/include/rclcpp/executor.hpp

```diff
@@ -548,7 +548,7 @@ class Executor
   spin_once_impl(std::chrono::nanoseconds timeout);
 
   typedef std::map<rclcpp::node_interfaces::NodeBaseInterface::WeakPtr,
-      const rcl_guard_condition_t *,
+      const rclcpp::GuardCondition *,
```

Comment:
As long as we're breaking API, let's change this to a reference (where ever it makes sense):

```suggestion
      const rclcpp::GuardCondition &,
```

This prevents the possibility of passing a nullptr into these methods that took pointers to guard conditions in the past. It puts the responsibility onto the caller of the methods to check this.

I think this is way more idiomatic C++, and I'd prefer to do it now, as we upgrade to C++ classes from C structs, instead of doing it in two steps, breaking API each time.

Reply:
With my last commit, the API is not being broken anymore. Anyway I tried this change, but having a reference in a map seems not behave nicely, as references must be init directly and can't be copied, I was getting a bunch of strange errors.

Reply:
Ah sorry, maybe the map should hold the pointer still, but in places where we pass by pointer, we could pass by reference instead (only in function/method interfaces).

Reply:
> the API is not being broken anymore

Also, we are still breaking API since we're changing the type from `rcl_guard_condition_t` to `rclcpp::GuardCondition` in several places in the API. Which is fine, but anyway it is breaking.

Reply:
Done in ccc78c5